site/js/Socket.js

```diff
@@ -0,0 +1,100 @@
+// socket.io-like wrapper around the native window.WebSocket.
+
+export default class Socket {
+  constructor(url) {
+    this.handlers = {}
+    this.url = url
+    this.isDead = false
+
+    this.reconnect()
+  }
+
+  on(evt, callbackFn) {
+    this.handlers[evt] = this.handlers[evt] || []
+    this.handlers[evt].push(callbackFn)
+  }
+
+  // Note: socket.io calls this `emit`.
+  send(evt, data) {
+    if (evt !== 'ping') {
+      console.info('socket.send[' + evt + ']::', data)
+    }
+
+    this.ws.send(JSON.stringify({
+      evt, data,
+    }))
+  }
+
+  diedHandler() {
+    if (this.isDead) {
+      return
+    }
+
+    this.isDead = true
+    console.error('socket:: is dead')
+
+    // TODO:
+    // Display some kind of onscreen notification that
+    // the connection to the server has been lost.
+
+    // Attempt to reconnect every couple seconds.
+    const intervalID = setInterval(() => {
+      if (!this.isDead) {
+        // We're alive now, no need to try again.
+        clearInterval(intervalID)
+
+        return
+      }
+
+      this.reconnect().then(() => {
+        // Connection re-established! :tada:
+        // (reconnect() automatically sets this.isDead = false)
+        console.info('socket:: connection re-established after death')
+      })
+    }, 2000)
+  }
```

Comment:
This whole automatic reconnect setup is super handy, thanks.

Reply:
It took ages to do, and loads of different ways to do it.

Originally the client pingponged the server (as the server does to the client), but that proved kinda buggy.